File: Binance-WebSoccket/rsibot/utils/extract_all_xpath_to_file-3.js
const fs = require("fs");
const { JSDOM } = require("jsdom");

function saveHTML(url, filename) {
  try {
    const htmlContent = fs.readFileSync(filename, "utf-8");
    console.log(`HTML content read from ${filename}`);
    return htmlContent;
  } catch (error) {
    console.error("Failed to read HTML file:", error);
    return null;
  }
}

function extractXPath(htmlContent) {
  try {
    const { window } = new JSDOM(htmlContent);
    const document = window.document;
    const elements = document.querySelectorAll("*");
    const xpathsWithIndex = Array.from(elements).map((element) =>
      getXPathWithIndex(element)
    );
    return xpathsWithIndex;
  } catch (error) {
    console.error("Failed to extract XPath:", error);
    return [];
  }
}

function getXPathWithIndex(element) {
  let xpath = [];
  for (let e = element; e; e = e.parentElement) {
    xpath.unshift(e.tagName.toLowerCase());
  }
  return "/" + xpath.map((tag, idx) => `${tag}[${idx + 1}]`).join("/");
}

function main() {
  const filename = "fnz_contact.html";
  const xpathFilename = "xpath-javascript-found-3.txt";

  const htmlContent = saveHTML(null, filename);
  if (!htmlContent) {
    console.log("HTML content not available.");
    return;
  }

  const xpathsWithIndex = extractXPath(htmlContent);

  console.log(
    "\nXPaths extracted from the HTML page with corresponding indices:"
  );
  xpathsWithIndex.forEach((xpath) => console.log(xpath));

  // Save the extracted XPaths with index to a file
  fs.writeFileSync(xpathFilename, xpathsWithIndex.join("\n"), "utf-8");
  console.log(`XPaths with corresponding indices saved to ${xpathFilename}`);
}

main();

File: Binance-WebSoccket/rsibot/utils/extract_all_xpath_to_file-3.py
import requests
from lxml import html

def save_html(url, filename):
    try:
        response = requests.get(url, verify=False)
        if response.status_code == 200:
            with open(filename, 'w', encoding='utf-8') as f:
                f.write(response.text)
            print(f"HTML page saved as {filename}")
        else:
            print("Failed to download HTML page")
    except Exception as e:
        print("Failed to download HTML page:", e)

def extract_xpath(html_file):
    try:
        with open(html_file, 'r', encoding='utf-8') as f:
            tree = html.fromstring(f.read())
            xpath_exprs = tree.xpath('//*')  # Get all elements
            xpaths_with_index = []
            for element in xpath_exprs:
                xpath = get_xpath_with_index(element)
                xpaths_with_index.append(xpath)
            return xpaths_with_index
    except Exception as e:
        print("Failed to extract XPath:", e)
        return []

def get_xpath_with_index(element):
    xpath = []
    for e in element.iterancestors():
        xpath.insert(0, e.tag)
    return '/'.join([f"{tag}[{idx}]" for idx, tag in enumerate(xpath, start=1)])

def main():
    url = "https://www.fnz.com/contact"
    filename = "fnz_contact.html"
    xpath_filename = "xpath-python-found-3.txt"

    save_html(url, filename)
    xpaths_with_index = extract_xpath(filename)

    print("\nXPaths extracted from the HTML page with corresponding indices:")
    for xpath in xpaths_with_index:
        print(xpath)

    # Save the extracted XPaths with index to a file
    with open(xpath_filename, 'w', encoding='utf-8') as f:
        for xpath in xpaths_with_index:
            f.write(f"{xpath}\n")
    print(f"XPaths with corresponding indices saved to {xpath_filename}")

if __name__ == "__main__":
    main()
